site/docs/coverage.md

```diff
@@ -0,0 +1,115 @@
+- A correct "instrumentation" configuration
+
+The former two are language-specific and mostly straightforward,
+however the latter can be more difficult for complex projects.
+
+"Instrumentation" in this case refers to the coverage tools that are
+used for a specific target. Bazel allows turning this on for a
+specific subset of files using the
+[`--instrumentation_filter`](../command-line-reference.html#flag--instrumentation_filter)
+flag, which specifies a filter for targets that are tested with the
+instrumentation enabled. To enable instrumentation for tests, the
+[`--instrument_test_targets`](../command-line-reference.html#flag--instrument_test_targets)
+is required.
+
+By default, bazel tries to match the target package(s), and prints the
+relevant filter as an `INFO` message.
+
+### Running coverage
+
+To produce a coverage report, use [`bazel coverage
+--combined_report=lcov
+[target]`](../command-line-reference.html#coverage). This runs the
+tests for the target, generating coverage reports in the lcov format
+for each file.
+
+Once finished, bazel runs an action that collects all the produced
+coverage files, and merges them into one, which is then finally
+created under `$(bazel info
+output_path)/_coverage/_coverage_report.dat`.
+
+Coverage reports are also produced if tests fail, though note that
+this does not extend to the failed tests - only passing tests are
+reported.
+
+### Viewing coverage
+
+The coverage report is only output in the non-human-readable `lcov`
+format. From this, we can use the `genhtml` utility (part of [the lcov
+project][lcov]) to produce a report that can be viewed in a web
+browser:
+
+```console
+genhtml --output genhtml "$(bazel info output_path)/_coverage/_coverage_report.dat"
+```
+
+Note that `genhtml` reads the source code as well, to annotate missing
+coverage in these files. For this to work, it is expected that
+`genhtml` is executed in the root of the bazel project.
+
+To view the result, simply open the `index.html` file produced in the
+`genhtml` directory in any web browser.
+
+For further help and information around the `genhtml` tool, or the
+`lcov` coverage format, see [the lcov project][lcov].
+
+## Remote execution
+
+Running with remote test execution currently has a few caveats:
+
+- The report combination action cannot yet run remotely. This is
+  because Bazel does not consider the coverage output files as part of
+  its graph, and can therefore not correctly treat them as inputs to
+  the combination action. To work around this, use
+  `--strategy=CoverageReport=local,remote`.
+- `--remote_download_minimal` and similar flags can also not be used
+  as a consequence of the former.
+- Bazel will currently fail to create coverage information if tests
+  have been cached previously. To work around this,
+  `--nocache_test_results` can be set specifically for coverage runs,
+  although this of course incurs a heavy cost in terms of test times.
```

Comment:
Do we have upstream issues for each of these items with an example? I think it'd be great if we can link them here.

Reply:
I've linked an issue for the `--strategy` flag, but haven't found one for `--nocache_test_results`

Reply:
Could you file the missing issues? Or link the repro here so that I can file an upstream issue for the ones that don't exist?

Reply:
Bradley filed a new issue for `--remote_download*`:

- https://github.com/bazelbuild/bazel/issues/14337
  - https://github.com/bazelbuild/bazel/issues/12592 also exists, but the above supersedes that because it's a full crash

We probably want to make an upstream issue for https://github.com/tweag/rules_haskell/issues/1613, I'm unsure if that's supposed to be https://github.com/bazelbuild/bazel/issues/12592.

Reply:
I suspect the caching issue may not have anything to do with remote_download issues, so better to file a separate issue I think

Reply:
I've noted this down separately in EngFlow/kotlin-perf-analysis#121, since the caching will be the first thing I look at I'll raise it as part of that soon.